ethers/providers/jsonrpc/subscriptions.nim

```diff
@@ -152,47 +150,60 @@ proc new*(_: type JsonRpcSubscriptions,
-      if "filter not found" in e.msg:
-        var newId: JsonNode
-        # Log filters are stored in logFilters, block filters are not persisted
-        # there is they do not need any specific data for their recreation.
-        # We use this to determine if the filter was log or block filter here.
-        if subscriptions.logFilters.hasKey(originalId):
-          let filter = subscriptions.logFilters[originalId]
-          newId = await subscriptions.client.eth_newFilter(filter)
-        else:
-          newId = await subscriptions.client.eth_newBlockFilter()
-        subscriptions.subscriptionMapping[originalId] = newId
-        return await getChanges(originalId)
+      var newId: JsonNode
+      # Log filters are stored in logFilters, block filters are not persisted
+      # there is they do not need any specific data for their recreation.
+      # We use this to determine if the filter was log or block filter here.
+      if subscriptions.logFilters.hasKey(id):
+        let filter = subscriptions.logFilters[id]
+        newId = await subscriptions.client.eth_newFilter(filter)
       else:
-        raise e
+        newId = await subscriptions.client.eth_newBlockFilter()
+      subscriptions.subscriptionMapping[id] = newId
+    except CancelledError as error:
+      raise error
+    except CatchableError:
+      # there's nothing further we can do here
+      discard
 
-  proc poll(id: JsonNode) {.async.} =
+  proc getChanges(id: JsonNode): Future[JsonNode] {.async: (raises: [CancelledError]).} =
+    if mappedId =? subscriptions.subscriptionMapping.?[id]:
+      try:
+        let changes = await subscriptions.client.eth_getFilterChanges(mappedId)
+        if changes.kind == JArray:
+          return changes
+      except JsonRpcError:
+        await resubscribe(id)
+        # TODO: we could still miss some events between losing the subscription
+        # and resubscribing. We should probably adopt a strategy like ethers.js,
+        # whereby we keep track of the latest block number that we've seen
+        # filter changes for:
+        # https://github.com/ethers-io/ethers.js/blob/f97b92bbb1bde22fcc44100af78d7f31602863ab/packages/providers/src.ts/base-provider.ts#L977
+      except CancelledError as error:
+        raise error
+      except CatchableError:
+        # there's nothing we can do here
+        discard
+    return newJArray()
+
+  proc poll(id: JsonNode) {.async: (raises: [CancelledError]).} =
     for change in await getChanges(id):
       if callback =? subscriptions.getCallback(id):
         callback(id, change)
 
-  proc poll {.async.} =
-    untilCancelled:
-      for id in toSeq subscriptions.callbacks.keys:
-        await poll(id)
-      await sleepAsync(pollingInterval)
+  proc poll {.async: (raises: []).} =
+    try:
+      while true:
+        for id in toSeq subscriptions.callbacks.keys:
+          await poll(id)
+        await sleepAsync(pollingInterval)
+    except CancelledError:
+      discard
```

Comment:
Normally we shouldn't swallow a CancelledError, but let it bubble up. But in this case we asyncSpawn the poll() future, and that means that any Error that bubbles up becomes a Defect and crashes the process. So that's why I catch the CancelledError outside of the loop, to make sure that the loop stops, and then I do not re-raise it because there is nothing that calls poll() that needs to be cancelled, and I don't want it to bubble up to asyncSpawn.